### PR TITLE
Default `benchmark` based on `deterministic` flag

### LINKED ISSUE #9128: [RFC] Depreceate `Trainer(benchmark)` and use deterministic
Out of scope for this PR but maybe we should remove `Trainer(benchmark)` and set

`torch.backends.cudnn.benchmark = not deterministic`

_Originally posted by @carmocca in https://github.com/PyTorchLightning/pytorch-lightning/pull/9121#discussion_r696202228_

cc @borda @justusschock @awaelchli @akihironitta @rohitgr7 @tchaton

### PR DESCRIPTION
**Pull request by krshrimali**
## What does this PR do?

* This PR makes the `benchmark` flag of `Trainer` class `Optional`, and defaults to the value based on `deterministic` flag. Also allows manually overwriting the flag.
* Why not deprecate? We wish to allow overwriting the flag by the users, and prevent BC-breaking changes.
* Tests have been updated accordingly.

Fixes #9128.

### Does your PR introduce any breaking changes? If yes, please list them.

No!

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Thanks to @carmocca for helping me with this PR. 🎉

### CODE REVIEW

**Review by carmocca — APPROVED** [review]
LGTM!

I pushed one commit with docs changes after searching for how to properly reference another parameter. Also added one reference to PyTorch's docs about randomness.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ def __init__(
         self.ipus = ipus
         self.num_nodes = num_nodes
         self.sync_batchnorm = sync_batchnorm
-        self.benchmark = benchmark
+        self.benchmark = not deterministic if benchmark is None else benchmark
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:125]: should this raise a misconfiguration error if someone explicitly specifies `benchmark=True, deterministic=True`?

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -637,23 +637,26 @@ def test_trainer_max_steps_accumulate_batches(tmpdir):
...
-    assert not torch.backends.cudnn.benchmark
+    original_val = torch.backends.cudnn.benchmark
 
-    # fit model
-    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, benchmark=True)
-    trainer.fit(model)
-
-    # verify training completed
-    assert trainer.state.finished, f"Training failed with {trainer.state}"
+    trainer = Trainer(benchmark=benchmark, deterministic=deterministic)
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:656]: This should silence the pre-commit error in CI

```suggestion
    _ = Trainer(benchmark=benchmark, deterministic=deterministic)
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:125]: IMO no because if `torch` allows it, we should too.

This PR is more about "better defaults" than constraining the available options

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Make `benchmark` flag optional, based on deterministic flag. ([#11944](https://github.com/PyTorchLightning/pytorch-lightning/pull/11944))
```
Comment (ananthsub) [inline CHANGELOG.md:12]: This should go into the Changed section of the changelog

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:125]: would this setting lead to a runtime error though? and if it does, shouldn't we either fail fast here, or indicate that `deterministic` has higher priority than `benchmark`, and override `benchmark` to be False?

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 1 above)
Comment (krshrimali) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:125]: > would this setting lead to a runtime error though? 

Can you please give a case where you think a runtime error might be raised? I can't think of such a case right now, hence the question.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -94,7 +94,7 @@ def __init__(
         amp_type: str = "native",
         amp_level: Optional[str] = None,
         sync_batchnorm: bool = False,
-        benchmark: bool = False,
+        benchmark: Optional[bool] = None,
         replace_sampler_ddp: bool = True,
         deterministic: bool = False,
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:99]: We currently don't do anything with the Trainer(deterministic) flag. The AcceleratorConnector rewrite #11448 missed this. I had a review comment but it was missed. 

We need to actually set the deterministic flag on the torch library / cudnn library. @krshrimali could you take care of it?

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -637,23 +637,26 @@ def test_trainer_max_steps_accumulate_batches(tmpdir):
...
+    [
+        (None, False, True),
+        (None, True, False),
+        (True, False, True),
+        (True, True, True),
+        (False, True, False),
+        (False, False, False),
+    ],
+)
+def test_benchmark_option(benchmark, deterministic, expected):
```
[the remark this answers is not in the document]
Comment (krshrimali) [reply tests/trainer/test_trainer.py:651]: Thanks, @awaelchli - I've taken a note of this. :)

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 6]
```diff
@@ -94,7 +94,7 @@ def __init__(
         amp_type: str = "native",
         amp_level: Optional[str] = None,
         sync_batchnorm: bool = False,
-        benchmark: bool = False,
+        benchmark: Optional[bool] = None,
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:97]: quick ques: why not deprecate this? we can keep this for next 2 release cycles and then remove it.

Comment (krshrimali) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:101]: Thanks, @rohitgr7 for the question.

If you meant deprecating the `benchmark` parameter, the reason is that we want to allow users to overwrite this flag manually, having a `benchmark` flag is quite intuitive in my opinion, and must be used a lot. Deprecation of this parameter might be a good discussion for a new issue if you think it's a valid direction to think about.

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:101]: I was just checking the issue, and that's what is proposed there. do you know what benefits do we have when both of these values are the same?

Comment (krshrimali) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:101]: Please see a related comment by @carmocca here: https://github.com/PyTorchLightning/pytorch-lightning/pull/11944/files#r808365722 on `benchmark=True, deterministic=True`. I am not sure about the benefits, I think if the user specifies the flags manually, those should be used. cc-ing: @carmocca who might be able to add something here.

Regarding deprecating this flag, I think `torch.backends.cudnn.deterministic` and `torch.backends.cudnn.benchmark` flags do two different things, and since PyTorch allows you to set such flags, it's probably a good idea to have them in the `Trainer` constructor. I'm not sure if PyTorch raises a warning (it doesn't on my machine), but lots of discussions happened [here](https://github.com/pytorch/pytorch/issues/6351).

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:101]: as discussed on the call let's not deprecate it and add a comment so we can look at this in future.

File: tests/trainer/test_trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/trainer/test_trainer.py:651]: #11999 has been merged, we can move forward with this. awesome!

**Review by awaelchli — APPROVED** [review]
GTLM

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 6 above)
Comment (krshrimali) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:102]: A warning has been added + refined test.

File: docs/source/common/trainer.rst [hunk 7]
```diff
@@ -416,18 +416,20 @@ benchmark
...
+Defaults to ``True`` if :paramref:`~pytorch_lightning.trainer.Trainer.deterministic` is not set.
+This flag sets the ``torch.backends.cudnn.deterministic`` flag. You can read more about its impact
+`here <https://pytorch.org/docs/stable/notes/randomness.html#cuda-convolution-benchmarking>`__
 
-The speedup comes from allowing the cudnn auto-tuner to find the best
-algorithm for the hardware `[see discussion here]
-<https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936>`_.
+This is likely to increase the speed of your system if your input sizes don't change. However, if they do, then it
+might make your system slower. The CUDNN auto-tuner will try to find the best algorithm for the hardware when a new
+input size is encountered, `ref <https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936>`__.
```
Comment (rohitgr7) [inline docs/source/common/trainer.rst:425]:
```suggestion
input size is encountered. Read more about it `here <https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936>`_.
```

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -228,8 +228,9 @@ def __init__(
                 that only one process at a time can access them.
                 Default: ``False``.
 
-            benchmark: If ``True``, enables cudnn.benchmark.
-                Default: ``False``.
+            benchmark: Sets ``torch.backends.cudnn.benchmark``.
+                Defaults to ``True`` if :paramref:`~pytorch_lightning.trainer.Trainer.deterministic` is ``False``.
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:232]:
```suggestion
                Defaults to ``True`` if :paramref:`~pytorch_lightning.trainer.trainer.Trainer.deterministic` is ``False``.
```

File: tests/trainer/test_trainer.py [hunk 9]
```diff
@@ -638,23 +638,30 @@ def test_trainer_max_steps_accumulate_batches(tmpdir):
...
-    trainer.fit(model)
-
-    # verify training completed
-    assert trainer.state.finished, f"Training failed with {trainer.state}"
+    if benchmark and deterministic:
+        with pytest.warns(RuntimeWarning, match="You passed `deterministic=True` and `benchmark=True`"):
+            _ = Trainer(benchmark=benchmark, deterministic=deterministic)
+    else:
+        _ = Trainer(benchmark=benchmark, deterministic=deterministic)
+    assert torch.backends.cudnn.benchmark == expected
```
Comment (rohitgr7) [inline tests/trainer/test_trainer.py:662]:
```suggestion
            trainer = Trainer(benchmark=benchmark, deterministic=deterministic)
    else:
        trainer = Trainer(benchmark=benchmark, deterministic=deterministic)
    assert torch.backends.cudnn.benchmark == expected
    assert trainer.benchmark == expected
```

**Review by rohitgr7 — APPROVED** [review]
LGTM
small comments.

File: tests/trainer/test_trainer.py [hunk 10]
```diff
@@ -638,23 +638,31 @@ def test_trainer_max_steps_accumulate_batches(tmpdir):
...
-
-    # verify training completed
-    assert trainer.state.finished, f"Training failed with {trainer.state}"
+    if benchmark and deterministic:
+        with pytest.warns(RuntimeWarning, match="You passed `deterministic=True` and `benchmark=True`"):
+            trainer = Trainer(benchmark=benchmark, deterministic=deterministic)
+    else:
+        trainer = Trainer(benchmark=benchmark, deterministic=deterministic)
+    assert torch.backends.cudnn.benchmark == expected
+    assert trainer.benchmark == expected
```
Comment (rohitgr7) [inline tests/trainer/test_trainer.py:663]:
```suggestion
    assert trainer._accelerator_connector.benchmark == expected
```
oh! it's accelerator connector.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 11]
```diff
@@ -142,8 +142,16 @@ def __init__(
             B. Strategy > Accelerator/precision/plugins
             C. TODO When multiple flag set to the same thing
         """
+        # Raise a warning when benchmark and deterministic are True
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:145]: nit: This comment does not provide any additional info or context

```suggestion
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 12]
```diff
@@ -142,8 +142,16 @@ def __init__(
...
+        if benchmark and deterministic:
+            rank_zero_warn(
+                "You passed `deterministic=True` and `benchmark=True`. Note that PyTorch ignores"
+                " torch.backends.cudnn.deterministic=True when torch.backends.cudnn.benchmark=True.",
+                category=RuntimeWarning,
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:150]: Why override the default `UserWarning`? This will only be shown after an user action, meaning, setting both of these flags

File: docs/source/common/trainer.rst [hunk 13]
```diff
@@ -416,18 +416,20 @@ benchmark
...
+Defaults to ``True`` if :paramref:`~pytorch_lightning.trainer.Trainer.deterministic` is not set.
+This flag sets the ``torch.backends.cudnn.deterministic`` flag. You can read more about its impact
+`here <https://pytorch.org/docs/stable/notes/randomness.html#cuda-convolution-benchmarking>`__
 
-The speedup comes from allowing the cudnn auto-tuner to find the best
-algorithm for the hardware `[see discussion here]
-<https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936>`_.
+This is likely to increase the speed of your system if your input sizes don't change. However, if they do, then it
+might make your system slower. The CUDNN auto-tuner will try to find the best algorithm for the hardware when a new
+input size is encountered. Read more about it `here <https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936>`_.
```
Comment (carmocca) [inline docs/source/common/trainer.rst:425]: This should use two underscores

```suggestion
input size is encountered. Read more about it `here <https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936>`__.
```

**Review by carmocca — APPROVED** [review]

File: docs/source/common/trainer.rst (hunk 13 above)
Comment (rohitgr7) [reply docs/source/common/trainer.rst]: quick ques: both are fine I think, but just curious is there any advantage of one over the other?

Comment (carmocca) [reply docs/source/common/trainer.rst]: since "here" is commonly used as the link text, using a single underscore could give clashes. I would suggest always using two.

File: tests/trainer/test_trainer.py [hunk 14]
```diff
@@ -638,23 +638,31 @@ def test_trainer_max_steps_accumulate_batches(tmpdir):
...
+    [
+        (None, False, True),
+        (None, True, False),
+        (True, False, True),
+        (True, True, True),
+        (False, True, False),
+        (False, False, False),
+    ],
+)
+def test_benchmark_option(benchmark, deterministic, expected):
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:652]: I think this should fix the pytest bug...

```suggestion
def test_benchmark_option(benchmark_, deterministic, expected):
```

File: tests/trainer/test_trainer.py [hunk 15]
```diff
@@ -638,23 +638,31 @@ def test_trainer_max_steps_accumulate_batches(tmpdir):
     assert trainer.global_step == trainer.max_steps, "Model did not stop at max_steps"
 
 
-def test_benchmark_option(tmpdir):
+@pytest.mark.parametrize(
+    ["benchmark", "deterministic", "expected"],
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:642]:
```suggestion
    ["benchmark_", "deterministic", "expected"],
```

File: tests/trainer/test_trainer.py [hunk 16]
```diff
@@ -638,23 +638,31 @@ def test_trainer_max_steps_accumulate_batches(tmpdir):
...
-    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, benchmark=True)
-    trainer.fit(model)
-
-    # verify training completed
-    assert trainer.state.finished, f"Training failed with {trainer.state}"
+    if benchmark and deterministic:
+        with pytest.warns(UserWarning, match="You passed `deterministic=True` and `benchmark=True`"):
+            trainer = Trainer(benchmark=benchmark, deterministic=deterministic)
+    else:
+        trainer = Trainer(benchmark=benchmark, deterministic=deterministic)
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:661]:
```suggestion
    if benchmark_ and deterministic:
        with pytest.warns(UserWarning, match="You passed `deterministic=True` and `benchmark=True`"):
            trainer = Trainer(benchmark=benchmark_, deterministic=deterministic)
    else:
        trainer = Trainer(benchmark=benchmark_, deterministic=deterministic)
```
